---
title: 'Helm chart config'
description: A reference of config fields for the Loculus Helm chart
---

import SchemaDocs from '../../../components/SchemaDocs.astro';

:::note
The config structure has not been finalized and may change rapidly until Loculus 1.0 is officially released (see [current state and roadmap](../../introduction/current-state-and-roadmap/)). In the meantime, this reference might also not be fully accurate or complete.
:::

The configuration for the Helm chart is provided as a YAML file. It has the following fields:

## General Settings

<SchemaDocs schemaPath='../kubernetes/loculus/values.schema.json' group='general' />

<div className='overflow-x-scroll'>
    <table className='min-w-[700px]'>
        <thead>
        <tr>
            <th className='w-56'>Field</th>
            <th>Type</th>
            <th className='w-32'>Default</th>
            <th>Description</th>
        </tr>
        </thead>
        <tbody>
        <tr>
            <td>`dataUseTerms.enabled`</td>
            <td>Boolean</td>
            <td>true</td>
            <td>Whether this Loculus instance handles data use terms.</td>
        </tr>
        <tr>
            <td>`dataUseTerms.urls.open`</td>
            <td>String</td>
            <td></td>
            <td>A URL describing the open data use terms.</td>
        </tr>
        <tr>
            <td>`dataUseTerms.urls.restricted`</td>
            <td>String</td>
            <td></td>
            <td>A URL describing the restricted data use terms.</td>
        </tr>
        </tbody>
    </table>
</div>

## Website Settings

<SchemaDocs schemaPath='../kubernetes/loculus/values.schema.json' group='website' />

## User registration and authentication

<SchemaDocs schemaPath='../kubernetes/loculus/values.schema.json' group='auth' />

## Database deployments

<table>
    <thead>
        <tr>
            <th className='w-72'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`runDevelopmentMainDatabase`</td>
            <td>Boolean</td>
            <td>true</td>
            <td>If true, runs a development database within the cluster.</td>
        </tr>
        <tr>
            <td>`runDevelopmentKeycloakDatabase`</td>
            <td>Boolean</td>
            <td>true</td>
            <td>If true, runs a development Keycloak database within the cluster.</td>
        </tr>
        <tr>
            <td>`developmentDatabasePersistence`</td>
            <td>Boolean</td>
            <td>true</td>
            <td>If true, makes the database on the argocd preview persistent.</td>
        </tr>
    </tbody>
</table>

For production environments, these should always be set to false. Instead, external managed databases should be used.

## Services

<table>
    <thead>
        <tr>
            <th className='w-48'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`disableWebsite`</td>
            <td>Boolean</td>
            <td>false</td>
            <td>If true, disables the frontend website deployment.</td>
        </tr>
        <tr>
            <td>`disableBackend`</td>
            <td>Boolean</td>
            <td>false</td>
            <td>If true, disables the backend API deployment.</td>
        </tr>
        <tr>
            <td>`disablePreprocessing`</td>
            <td>Boolean</td>
            <td>false</td>
            <td>If true, disables preprocessing pipelines.</td>
        </tr>
        <tr>
            <td>`disableIngest`</td>
            <td>Boolean</td>
            <td>false</td>
            <td>If true, disables ingestion services.</td>
        </tr>
        <tr>
            <td>`disableEnaSubmission`</td>
            <td>Boolean</td>
            <td>false</td>
            <td>If true, disables ENA (European Nucleotide Archive) submission service.</td>
        </tr>
        <tr>
            <td>`customWebsiteImage`</td>
            <td>String</td>
            <td></td>
            <td>Optional custom Docker image for the website.</td>
        </tr>
    </tbody>
</table>

## Organism Configuration

<table>
    <thead>
        <tr>
            <th className='w-48'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`organisms`</td>
            <td>Object</td>
            <td></td>
            <td>An object where the keys are the organism IDs and values are an [Organism (type)](#organism-type)</td>
        </tr>
        <tr>
            <td>`lineageSystemDefinitions`</td>
            <td>Object</td>
            <td></td>
            <td>An object where the keys are the lineage system names and values are links to lineage system definition files per pipeline version (See [Lineage system defintions](#lineage-type))</td>

        </tr>
    </tbody>

</table>

### Lineage system definitions

Here's an example of a `lineageDefinitions` section:

```yaml
lineageSystemDefinitions:
    pangoLineage: # Lineage name to use in metadata fields
        1: https://example.org/lineage_defintions_v1.yaml # Definition per pipeline version
        2: https://example.org/lineage_defintions_v2.yaml
    myLineage:
        1: ...
```

<table>
    <thead>
        <tr>
            <th className='w-48'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`lineageSystemDefinitions.<name>`</td>
            <td>Object</td>
            <td></td>
            <td>A map from pipeline versions to file URLs.</td>
        </tr>
        <tr>
            <td>`lineageSystemDefinitions.<name>.<pipelineVersion>`</td>
            <td>`String`</td>
            <td></td>
            <td>The URL to the lineage defintion file for that lineage system and that pipeline version.</td>
        </tr>
    </tbody>
</table>

### Organism (type)

Each organism object has the following fields:

<table>
    <thead>
        <tr>
            <th className='w-48'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`schema`</td>
            <td>[Schema (type)](#schema-type)</td>
            <td></td>
            <td></td>
        </tr>
        <tr>
            <td>`preprocessing`</td>
            <td>A list of [Preprocessing (type)](#preprocessing-type)</td>
            <td></td>
            <td>Configuration for the preprocessing pipeline(s)</td>
        </tr>
        <tr>
            <td>`ingest`</td>
            <td>[Ingest (type)](#ingest-type)</td>
            <td></td>
            <td>Configuration for the ingest pipeline(s)</td>
        </tr>
        <tr>
            <td>`referenceGenomes`</td>
            <td>Object</td>
            <td></td>
            <td>Reference genome sequences for the organism</td>
        </tr>
        <tr>
            <td>`referenceGenomes.nucleotideSequences`</td>
            <td>Array of [NucleotideSequence (type)](#nucleotidesequence-type)</td>
            <td></td>
            <td></td>
        </tr>
        <tr>
            <td>`referenceGenomes.genes`</td>
            <td>Array of [Gene (type)](#gene-type)</td>
            <td></td>
            <td></td>
        </tr>
    </tbody>
</table>

### Schema (type)

<table>
    <thead>
        <tr>
            <th className='w-60'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`organismName`</td>
            <td>String</td>
            <td></td>
            <td>Display name for the organism</td>
        </tr>
        <tr>
            <td>`image`</td>
            <td>String</td>
            <td></td>
            <td>URL to an image that will be shown on the landing page</td>
        </tr>
        <tr>
            <td>`loadSequencesAutomatically`</td>
            <td>Boolean</td>
            <td>true</td>
            <td>
                Whether sequences be loaded automatically on the sequence details page , rather than users having to
                press a button to do so. (For small genomes, this should probably be true.)
            </td>
        </tr>
        <tr>
            <td>`submissionDataTypes.consensusSequences`</td>
            <td>Boolean</td>
            <td>true</td>
            <td>
                If `false`, the submission form will not allow submission of consensus sequences (i.e. the sequences
                file must be omitted). All consensus sequence related parts on the website will be hidden.
            </td>
        </tr>
        <tr>
            <td>`description`</td>
            <td>String</td>
            <td></td>
            <td></td>
        </tr>
        <tr>
            <td>`metadata`</td>
            <td>Array of [Metadata (type)](#metadata-type)</td>
            <td></td>
            <td>Metadata fields associated with the organism.</td>
        </tr>
        <tr>
            <td>`metadataTemplate`</td>
            <td>Array of String</td>
            <td></td>
            <td>
                Which input fields to add to the downloadable metadata template on the submission and revision page.
            </td>
        </tr>
        <tr>
            <td>`earliestReleaseDate`</td>
            <td>Object</td>
            <td></td>
            <td>
                Configuration object for enabling and configuring the `earliestReleaseDate` metadata field. For each
                version of an accession, the `earliestReleaseDate` is calculated as the earliest date of the internal
                release date, the dates in the configured `externalFields` and the value from the previous version of
                the accession (if there is one). This can be used when having a mix of sequences imported from other
                databases, as well as sequences released first in this Loculus instance, to have a field that shows the
                earliest release date regardless of where the sequence was first released.
            </td>
        </tr>
        <tr>
            <td>`earliestReleaseDate.enabled`</td>
            <td>boolean</td>
            <td></td>
            <td>Whether to enable the `earliestReleaseDate` metadata field.</td>
        </tr>
        <tr>
            <td>`earliestReleaseDate.externalFields`</td>
            <td>Array of strings</td>
            <td></td>
            <td>
                Field names to use when calculating the earliest release date. The fields need to be nullable strings
                formated with `yyyy-mm-dd`.
            </td>
        </tr>
        <tr>
            <td>`website`</td>
            <td>Object</td>
            <td></td>
            <td>Configuration for how the organism data is displayed on the website</td>
        </tr>
        <tr>
            <td>`website.tableColumns`</td>
            <td>Array of Strings</td>
            <td></td>
            <td>Columns to display in the browse table</td>
        </tr>
        <tr>
            <td>`website.defaultOrderBy`</td>
            <td>String</td>
            <td></td>
            <td>Default column to sort the browse table</td>
        </tr>
        <tr>
            <td>`website.defaultOrder`</td>
            <td>"ascending" | "descending"</td>
            <td></td>
            <td>Default order direction</td>
        </tr>
        <tr>
            <td>`silo`</td>
            <td>Object</td>
            <td></td>
            <td>Configuration regarding the SILO database engine</td>
        </tr>
        <tr>
            <td>`silo.dateToSortBy`</td>
            <td>String</td>
            <td></td>
            <td>
                Name of a field of type date. This might speed up date range searches on this field and improve sequence
                compression.
            </td>
        </tr>
        <tr>
            <td>`silo.partitionBy`</td>
            <td>String</td>
            <td></td>
            <td></td>
        </tr>
    </tbody>
</table>

### Metadata (type)

<table>
    <thead>
        <tr>
            <th className='w-56'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`name`</td>
            <td>String</td>
            <td></td>
            <td>Key used across app to refer to this field</td>
        </tr>
        <tr>
            <td>`displayName`</td>
            <td>String</td>
            <td></td>
            <td>Name displayed to users</td>
        </tr>
        <tr>
            <td>`type`</td>
            <td></td>
            <td></td>
            <td></td>
        </tr>
        <tr>
            <td>`header`</td>
            <td>String</td>
            <td></td>
            <td>Grouping of fields in sequence details UI</td>
        </tr>
        <tr>
            <td>`required`</td>
            <td>Boolean</td>
            <td></td>
            <td>Whether the field is required by backend</td>
        </tr>
        <tr>
            <td>`desired`</td>
            <td>Boolean</td>
            <td></td>
            <td>Whether the field is a desired input field for submitters</td>
        </tr>
        <tr>
            <td>`definition`</td>
            <td>Boolean</td>
            <td></td>
            <td>Definition of input field for submitters</td>
        </tr>
        <tr>
            <td>`guidance`</td>
            <td>Boolean</td>
            <td></td>
            <td>Guidance for submitters on filling in input field</td>
        </tr>
        <tr>
            <td>`noInput`</td>
            <td>Boolean</td>
            <td></td>
            <td>
                Whether a field with this name is expected as possible input, and so should be included in the metadata
                template and as a form field. (If set to true it will not be included).
            </td>
        </tr>
        <tr>
            <td>`generateIndex`</td>
            <td>Boolean</td>
            <td></td>
            <td>
                Whether the field should be indexed for search. This is only allowed for string fields and facilitates
                faster filters. It is recommended if the number of different values is rather small.
            </td>
        </tr>
        <tr>
            <td>`columnWidth`</td>
            <td>Number</td>
            <td></td>
            <td>The minimum column width for this field on the search table.</td>
        </tr>
        <tr>
            <td>`order`</td>
            <td>Number</td>
            <td></td>
            <td>Order for the column in the search table, lower first</td>
        </tr>
        <tr>
            <td>`autocomplete`</td>
            <td>Boolean</td>
            <td></td>
            <td>
                Whether autocomplete should be offered for the field. This is only allowed for string fields and
                probably `generateIndex` should be true.
            </td>
        </tr>
        <tr>
            <td>`enableSubstringSearch`</td>
            <td>Boolean</td>
            <td></td>
            <td>
                If true, search results will contain results that contain the given value as a substring. If false (the
                default), only exact matches will be returned. This only works for string fields, and you cannot also
                enable `autocomplete`.
            </td>
        </tr>
        <tr>
            <td>`rangeSearch`</td>
            <td>Boolean</td>
            <td></td>
            <td>If true, enables range search for numeric fields.</td>
        </tr>
        <tr>
            <td>`rangeOverlapSearch`</td>
            <td>Object</td>
            <td></td>
            <td>The config settings for enabling range overlap search.</td>
        </tr>
        <tr>
            <td>`rangeOverlapSearch.rangeId`</td>
            <td>String</td>
            <td></td>
            <td>
                The range that this field belongs to. Two fields (the upper and lower bound) need to be defined with the
                same range ID.
            </td>
        </tr>
        <tr>
            <td>`rangeOverlapSearch.rangeDisplayName`</td>
            <td>String</td>
            <td></td>
            <td>The display name of the range.</td>
        </tr>
        <tr>
            <td>`rangeOverlapSearch.bound`</td>
            <td>"lower" | "upper"</td>
            <td></td>
            <td>Whether this field is the lower or upper bound of the range.</td>
        </tr>
        <tr>
            <td>`initiallyVisible`</td>
            <td>Boolean</td>
            <td></td>
            <td>If true, the field is initially visible in the UI.</td>
        </tr>
        <tr>
            <td>`hideOnSequenceDetailsPage`</td>
            <td>Boolean</td>
            <td></td>
            <td>If true, hides the field on the sequence details page.</td>
        </tr>
        <tr>
            <td>`hideInSearchResultsTable`</td>
            <td>Boolean</td>
            <td></td>
            <td>If true, hides the field in the search results table (and makes it impossible to show).</td>
        </tr>
        <tr>
            <td>`perSegment`</td>
            <td>Boolean</td>
            <td></td>
            <td>
                Whether this is a metadata field that should exist for each segment. If so fields will be created as
                `fieldName_A`, `fieldName_B` in the case of an organism with segments `A` and `B`.
            </td>
        </tr>
        <tr>
            <td>`customDisplay`</td>
            <td>Object</td>
            <td></td>
            <td>Custom display of the field on the sequence details page</td>
        </tr>
        <tr>
            <td>`customDisplay.type`</td>
            <td></td>
            <td></td>
            <td></td>
        </tr>
        <tr>
            <td>`customDisplay.url`</td>
            <td></td>
            <td></td>
            <td></td>
        </tr>
        <tr>
            <td>`ontology_id`</td>
            <td>String</td>
            <td></td>
            <td></td>
        </tr>
        <tr>
            <td>`definition`</td>
            <td>String</td>
            <td></td>
            <td></td>
        </tr>
        <tr>
            <td>`guidance`</td>
            <td>String</td>
            <td></td>
            <td></td>
        </tr>
        <tr>
            <td>`example`</td>
            <td>String</td>
            <td></td>
            <td>Example value for the field.</td>
        </tr>
        <tr>
            <td>`ingest`</td>
            <td>String</td>
            <td></td>
            <td>Which NCBI field to map to this field</td>
        </tr>
        <tr>
            <td>`preprocessing`</td>
            <td>Object</td>
            <td></td>
            <td>
                The values of this field will be added to the preprocessing pipeline config file and the available
                values depend on the chosen pipeline. For the Nextclade pipeline, please see
                [here](https://github.com/loculus-project/loculus/blob/main/preprocessing/nextclade/README.md#custom-preprocessing-functions).
            </td>
        </tr>
        <tr>
            <td>`lineageSystem`</td>
            <td>String</td>
            <td></td>
            <td>
                Use this on string fields that contain lineages, if you want to enable searches that can include
                sublineages. The value needs to be a lineage system that is defined under the `lineageSystemDefinitions`
                key.
            </td>
        </tr>
    </tbody>
</table>

### Preprocessing (type)

<table>
    <thead>
        <tr>
            <th className='w-48'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`version`</td>
            <td>Integer</td>
            <td></td>
            <td>Version of the preprocessing pipeline</td>
        </tr>
        <tr>
            <td>`image`</td>
            <td>String</td>
            <td></td>
            <td>Docker image for the preprocessing pipeline</td>
        </tr>
        <tr>
            <td>`args`</td>
            <td>Array of strings</td>
            <td></td>
            <td>Arguments passed to the preprocessing pipeline</td>
        </tr>
        <tr>
            <td>`configFile`</td>
            <td>Object</td>
            <td></td>
            <td>Fields that should be added to the preprocessing pipeline config file</td>
        </tr>
    </tbody>
</table>

The values for `args` and `configFile` depend on the used preprocessing pipeline. For the Nextclade preprocessing pipeline, please see [here](../../for-administrators/existing-preprocessing-pipelines/#nextclade-based-pipeline).

### Ingest (type)

<table>
    <thead>
        <tr>
            <th className='w-48'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`image`</td>
            <td>"ghcr.io/loculus-project/ingest"</td>
            <td></td>
            <td>Docker image for the ingest pipeline</td>
        </tr>
        <tr>
            <td>`configFile`</td>
            <td>Object</td>
            <td></td>
            <td>Fields that should be added to the ingest pipeline config file</td>
        </tr>
    </tbody>
</table>

The values for `configFile` depend on the used preprocessing pipeline. For [our ingest pipeline](https://github.com/loculus-project/loculus/tree/main/ingest) which downloads data from NCBI GenBank using NCBI Datasets, the config file needs to contain the taxon_id of the organism and additionally, if the organism is multi-segmented, it requires a list of segment names (nucleotide_sequences) and the nextclade_dataset that can be used for segment identification and alignment.

### NucleotideSequence (type)

<table>
    <thead>
        <tr>
            <th className='w-48'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`name`</td>
            <td>String</td>
            <td></td>
            <td>Name of the sequence</td>
        </tr>
        <tr>
            <td>`sequence`</td>
            <td></td>
            <td></td>
            <td></td>
        </tr>
        <tr>
            <td>`insdcAccessionFull`</td>
            <td>String</td>
            <td></td>
            <td>INSDC accession of the sequence</td>
        </tr>
    </tbody>
</table>

### Gene (type)

<table>
    <thead>
        <tr>
            <th className='w-48'>Field</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>`name`</td>
            <td>String</td>
            <td></td>
            <td>Name of the sequence</td>
        </tr>
        <tr>
            <td>`sequence`</td>
            <td></td>
            <td></td>
            <td></td>
        </tr>
    </tbody>
</table>
